skorch/hf.py

```diff
@@ -0,0 +1,800 @@
+"""Classes to work with huggingface libraries
+
+E.g. transformers or tokenizers
+
+This module should be treated as a leaf node in the dependency tree, i.e. no
+other skorch modules should depend on these classes or import from here. Even
+so, don't import any Huggingface libraries on the root level because skorch
+should not depend on them.
+
+"""
+
+import os
+from copy import deepcopy
+from operator import itemgetter
+
+import numpy as np
+from sklearn.base import BaseEstimator, TransformerMixin
+
+from skorch.utils import check_is_fitted, params_for
+
+
+class _HuggingfaceTokenizerBase(BaseEstimator, TransformerMixin):
+    """Base class for yet to train and pretrained tokenizers
+
+    Implements the ``vocabulary_`` attribute and the methods
+    ``get_feature_names``, ``transform``, and ``inverse_transform``.
+
+    Subclasses should implement the ``fit``  method.
+
+    """
+    @property
+    def vocabulary_(self):
+        check_is_fitted(self, ['fast_tokenizer_'])
+        return self.fast_tokenizer_.vocab
+
+    def get_feature_names(self):
+        """Array mapping from feature integer indices to feature name.
+
+        Note: Same implementation as sklearn's CountVectorizer
+
+        Returns
+        -------
+        feature_names : list
+            A list of feature names.
+
+        """
+        return [t for t, i in sorted(self.vocabulary_.items(), key=itemgetter(1))]
+
+    def fit(self, X, y=None, **fit_params):
+        raise NotImplementedError
+
+    def transform(self, X):
+        """Transform the given data
+
+        Parameters
+        ----------
+        X : iterable of str
+          A list/array of strings or an iterable which generates either strings.
+
+        Returns
+        -------
+        Xt : transformers.tokenization_utils_base.BatchEncoding
+          A huggingface ``BatchEncoding`` instance. This is basically a
+          dictionary containing the ids of the tokens and some additional fields
+          (depending on the parameters), e.g. the attention mask. For this
+          reason, the output is not well suited to be used with normal sklearn
+          models, but it works with huggingface transformers and with skorch
+          nets.
+
+        """
+        check_is_fitted(self, ['fast_tokenizer_'])
```

Comment:
you can also implement `__sklearn_is_fitted__` and call `check_is_fitted(self)` to do the same thing, and allow other estimators calling `check_is_fitted` on this estimator to do the same thing.

Reply:
[Here](https://scikit-learn.org/stable/modules/generated/sklearn.utils.validation.check_is_fitted.html), it says

> If an estimator does not set any attributes with a trailing underscore, it can define a __sklearn_is_fitted__ method returning a boolean to specify if the estimator is fitted or not.

But from your comment I take that another advantage of implementing that method is that it allows other sklearn code to check if the estimator is fitted without the caller having to know what is needed for an estimator to be fitted, right?

Reply:
Got it, I'll implement it this way.